### PR TITLE
Catchup with posix features

### PR DESCRIPTION
**Pull request by yuslepukhin**
Catch up with Posix features
  NewWritableRWFile must fail when file does not exists
  Implement Env::Truncate()
  Adjust Env options optimization functions
  Implement MemoryMappedBuffer on Windows.

### CODE REVIEW

File: include/rocksdb/env.h [hunk 1]
```diff
@@ -822,13 +822,24 @@ class RandomRWFile {
...
   MemoryMappedFileBuffer(void* _base, size_t _length)
-      : base(_base), length(_length) {}
+      : base_(_base), length_(_length) {}
+
   virtual ~MemoryMappedFileBuffer() = 0;
 
-  void* const base;
-  const size_t length;
+  // We do not want to unmap this twice. We can make this class
+  // movable if desired, however, since
```
Comment (ajkr) [inline include/rocksdb/env.h]: since only the moved-to object will hold the mapping? just guessing :)

File: port/win/io_win.h [hunk 2]
```diff
@@ -411,6 +411,18 @@ class WinRandomRWFile : private WinFileData,
   virtual Status Close() override;
 };
 
+class WinMemoryMappedBufer : public MemoryMappedFileBuffer {
```
Comment (ajkr) [inline port/win/io_win.h]: s/Bufer/Buffer/

File: include/rocksdb/env.h (hunk 1 above)
Comment (yuslepukhin) [reply include/rocksdb/env.h]: @ajkr Yes, just bc unique_ptr is used to hold it but inside the implementation nothing prevents copying. Also the buffer details are exposed.

File: port/win/io_win.h (hunk 2 above)
Comment (yuslepukhin) [reply port/win/io_win.h]: Will address shortly.

**Comment by yuslepukhin** [comment]
@miasantreble Do we want to merge it?

**Pull request closed without merging**